src/__experimental__/components/select-async/select-async.component.js

```diff
@@ -0,0 +1,185 @@
+ *
+ * The opts object is used with axios, please see this URL for more options:
+ * https://github.com/axios/axios#request-config
+ *
+ * # The Response
+ *
+ * By default the response is expected in the following format:
+ *
+ *   {
+ *     $items: [Array] - required, the items to render
+ *     $page: [Integer] - optional, for paginated responses
+ *     $total: [Integer] - optional, for paginated responses
+ *   }
+ *
+ * If your response does not match this format you can modify it using the onResponse prop:
+ *
+ *   <SelectAsync onResponse={ response => ({ $items: response.myItems }) } />`
+ *
+ * # Rendering Options
+ *
+ * By default options will be rendered under the assumption they have the following format:
+ *
+ *   {
+ *     id [Integer] - the unique identifier
+ *     displayed_as [String] - the human readable value
+ *   }
+ *
+ * If your items do not match this format, or you want to have custom control over how they
+ * are rendered, you can provide a function as a child:
+ *
+ *   <SelectAsync>
+ *     {
+ *       items => items.map(item => <Option>{ item.text }</Option>)
+ *     }
+ *   </SelectAsync>
+ */
+class SelectAsync extends React.Component {
+  static propTypes = {
+    children: PropTypes.func,
+    endpoint: PropTypes.string.isRequired,
+    onRequest: PropTypes.func,
+    onResponse: PropTypes.func
+  }
+
+  state = {
+    fetching: false,
+    items: [],
+    page: 0,
+    total: 0
+  }
+
+  select = React.createRef()
+
+  deferredFetch = undefined
+
+  onOpen = () => this.fetchData({ page: 1 });
+
+  filterValue() {
+    return this.select.current.state.filter;
+  }
+
+  fetchOptions(page, search) {
+    let opts = {
+      params: {
+        page,
+        items_per_page: 20,
+        search
+      }
+    };
+
+    if (this.props.onRequest) opts = this.props.onRequest(opts);
+
+    return opts;
+  }
+
+  handleResponse(responseData) {
+    let data = responseData;
+
+    if (this.props.onResponse) data = this.props.onResponse(data);
+
+    if (!Object.prototype.hasOwnProperty.call(data, '$items')) {
+      throw Error(responseErrorMessage);
+    }
+
+    this.setState({
+      items: this.buildOptions(data.$page, data.$items),
+      page: data.$page || 1,
+      total: data.$total || data.$items.length,
+      fetching: false
+    });
+  }
+
+  deferredFetchData = (search) => {
+    clearTimeout(this.deferredFetch);
+    this.deferredFetch = setTimeout(() => {
+      this.fetchData({ page: 1, search });
+    }, 200);
+  }
+
+  fetchData = async ({ page, search = this.filterValue() }) => {
+    this.setState({ fetching: true });
+    const { data } = await axios.get(
+      this.props.endpoint,
+      this.fetchOptions(page, search)
+    );
+    this.handleResponse(data);
+  }
+
+  fetchNextPage = () => {
+    if (this.state.fetching) return;
+    if (this.state.items.length === this.state.total) return;
+    this.fetchData({ page: this.state.page + 1 });
```

Comment:
object destructuring could be used here, to refer ```this.state``` only once

Reply:
✅